### PR TITLE
feat(cloud): add pagination for cloud `list` operations

### PR DESCRIPTION
**Pull request by aaronsteers**
## Summary
Adds optional `limit` and `offset` pagination support to PyAirbyte Cloud list operations for connections, workspaces, sources, and destinations, while preserving the existing auto-pagination behavior when no limit is provided.

Also adds pagination for `get_job_logs()` so sync history requests fetch additional pages when the requested limit exceeds one API page. The list operation implementation preserves exact empty-string `name` filtering, keeps filtered list fetches at full API page size before applying client-side result limits, rejects non-positive `limit` values, and distinguishes true job-log 404s from auth/service failures.

Removed stale pagination TODOs from workspace list wrappers and added focused unit coverage for job log pagination plus resource-list pagination edge cases. Added docstrings for the new pagination test helpers/tests after CodeRabbit flagged docstring coverage.

Requested by AJ Steers from Slack thread: https://airbytehq-team.slack.com/archives/C08BHPUMEPJ/p1779121915926609?thread_ts=1779121915.926609&cid=C08BHPUMEPJ

## Review & Testing Checklist for Human
- [ ] Verify the new `limit` and `offset` parameters on `CloudWorkspace.list_connections()`, `list_sources()`, and `list_destinations()` match expected public API semantics.
- [ ] Confirm filtered list calls with `name` or `name_filter` should continue scanning full API pages until enough matching results are found or the API is exhausted.
- [ ] Check that `limit=0` should raise `PyAirbyteInputError`, while `offset=0` remains valid.
- [ ] Check that `get_job_logs()` non-2xx handling matches caller expectations: 404 maps to `AirbyteMissingResourceError`, other non-2xx responses map to `AirbyteError`.
- [ ] Run an end-to-end Cloud workspace smoke test with a workspace containing more than one API page of jobs or resources, if available.

### Notes
Local validation completed:
- `uv run --project /home/ubuntu/repos/PyAirbyte ruff format --check /home/ubuntu/repos/PyAirbyte/airbyte/_util/api_util.py /home/ubuntu/repos/PyAirbyte/airbyte/cloud/workspaces.py /home/ubuntu/repos/PyAirbyte/tests/unit_tests/test_cloud_api_util.py`
- `uv run --project /home/ubuntu/repos/PyAirbyte ruff check /home/ubuntu/repos/PyAirbyte/airbyte/_util/api_util.py /home/ubuntu/repos/PyAirbyte/airbyte/cloud/workspaces.py /home/ubuntu/repos/PyAirbyte/tests/unit_tests/test_cloud_api_util.py`
- `uv run --project /home/ubuntu/repos/PyAirbyte ruff check /home/ubuntu/repos/PyAirbyte/tests/unit_tests/test_cloud_api_util.py --select D`
- `uv run --project /home/ubuntu/repos/PyAirbyte pyrefly check /home/ubuntu/repos/PyAirbyte/airbyte/_util/api_util.py /home/ubuntu/repos/PyAirbyte/airbyte/cloud/workspaces.py /home/ubuntu/repos/PyAirbyte/tests/unit_tests/test_cloud_api_util.py --output-format=json`
- `uv run --project /home/ubuntu/repos/PyAirbyte pytest /home/ubuntu/repos/PyAirbyte/tests/unit_tests/test_cloud_api_util.py -q`
- Real Airbyte Cloud smoke test against the Devin Sandbox workspace before the latest reviewer-only cleanup: 55 connections, 65 sources, 19 destinations, `list_connections(limit=2)`, and `get_previous_sync_logs(limit=1)` completed successfully.

---
[Devin session](https://app.devin.ai/sessions/c7675392e71e4d7087214b0274123bbb)

Requested by: @aaronsteers

## Summary by CodeRabbit

* **New Features**
  * List endpoints now support optional limit-based pagination for connections, workspaces, sources, and destinations; results are capped to the requested limit.
  * Job log retrieval auto-paginates across pages to fulfill larger limits.

* **Improvements**
  * Unified limit-style controls across cloud tools and simplified tailing default to from_tail=True.
  * Stronger input validation for pagination and mutual exclusion of name vs name_filter.

* **Tests**
  * Expanded unit tests covering pagination, filtering, and edge cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/airbytehq/PyAirbyte/pull/1024?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: airbyte/_util/api_util.py [hunk 1]
```diff
@@ -60,6 +57,32 @@ def status_ok(status_code: int) -> bool:
...
 
 
+def _validate_pagination_params(
+    *,
+    limit: int | None,
+    offset: int | None,
+) -> None:
+    """Validate common pagination parameters."""
+    if limit is not None and limit < 0:
+        raise PyAirbyteInputError(message="`limit` must be greater than or equal to 0.")
```
Comment (aaronsteers) [inline airbyte/_util/api_util.py]: Let's make 0 limit illegal. So offset can be 0 but limit but be greater than 0. (Otherwise undefined if the action should be a short-circuited no-op or a real call to the API.)

File: airbyte/_util/api_util.py [hunk 2]
```diff
@@ -250,7 +273,7 @@ def get_workspace(
 # List resources
 
 
-def list_connections(
+def list_connections(  # noqa: PLR0913
```
Comment (aaronsteers) [inline airbyte/_util/api_util.py]: You must comment with at least the name of the rule you are noqa-ing.

File: airbyte/_util/api_util.py [hunk 3]
```diff
@@ -419,12 +472,25 @@ def list_sources(
...
+        if not response.sources_response.next:
+            break
+
+        current_offset += len(page_data)
 
     return result
 
 
-def list_destinations(
+def list_destinations(  # noqa: PLR0913
```
Comment (aaronsteers) [inline airbyte/_util/api_util.py]: Ditto.

**Review by aaronsteers — COMMENTED** [review]
Let's simplify and remove support for `offset`. So we'll support `limit` only in our Python API. Also add `limit` support on the relevant MCP tools if it does not already exist.

File: airbyte/cloud/connections.py [hunk 4]
```diff
@@ -308,7 +308,6 @@ def get_previous_sync_logs(
         self,
         *,
         limit: int = 20,
-        offset: int | None = None,
```
Comment (aaronsteers) [inline airbyte/cloud/connections.py:311]: Bring back this option here, to avoid the breaking change. (No need to push down if inner function doesn't support.)

Comment (aaronsteers) [reply airbyte/cloud/connections.py:311]: To be clear: bring back whereever it existed prior to this PR. Just don't add unnecessarily everywhere (to avoid added complexity in this PR).
